libs/vkd3d/command.c

```diff
@@ -10118,6 +10174,174 @@ static void STDMETHODCALLTYPE d3d12_command_list_DiscardResource(d3d12_command_l
     }
 }
 
+static void d3d12_command_list_resolve_binary_occlusion_queries(struct d3d12_command_list *list,
+        VkBuffer src_buffer, uint32_t src_index, VkBuffer dst_buffer, VkDeviceSize dst_offset,
+        VkDeviceSize dst_size, uint32_t dst_index, uint32_t count);
+
+static uint32_t vkd3d_query_lookup_entry_hash(const void *key)
+{
+    const struct vkd3d_query_lookup_key *k = key;
+
+    /* The bucket index is expected to be small and may fit into the
+     * lower bits of the heap address, which are expected to be zero */
+    return ((uintptr_t)k->query_heap) ^ k->bucket;
+}
+
+static bool vkd3d_query_lookup_entry_compare(const void *key, const struct hash_map_entry *entry)
+{
+    const struct vkd3d_query_lookup_key *a = key;
+    const struct vkd3d_query_lookup_key *b = &((const struct vkd3d_query_lookup_entry*)entry)->key;
+
+    return a->query_heap == b->query_heap && a->bucket == b->bucket;
+}
+
+static bool d3d12_command_list_is_query_resolve_pending(struct d3d12_command_list *list,
+        struct d3d12_query_heap *query_heap, uint32_t query_index)
+{
+    struct vkd3d_query_lookup_entry *e;
+    struct vkd3d_query_lookup_key key;
+    uint32_t bit_index;
+
+    key.query_heap = query_heap;
+    key.bucket = query_index >> VKD3D_QUERY_LOOKUP_GRANULARITY_BITS;
+
+    e = (struct vkd3d_query_lookup_entry*)hash_map_find(&list->query_resolve_lut, &key);
+
+    if (!e)
+        return false;
+
+    bit_index = query_index & VKD3D_QUERY_LOOKUP_INDEX_MASK;
+    return !!(e->query_mask & (1ull << bit_index));
+}
+
+static void d3d12_command_list_add_query_lookup_mask(struct d3d12_command_list *list,
+        struct d3d12_query_heap *query_heap, uint32_t bucket, uint64_t query_mask)
+{
+    struct vkd3d_query_lookup_entry entry, *e;
+
+    entry.key.query_heap = query_heap;
+    entry.key.bucket = bucket;
+    entry.query_mask = 0ull;
+
+    e = (struct vkd3d_query_lookup_entry*)hash_map_insert(&list->query_resolve_lut, &entry.key, &entry.hash_entry);
+    e->query_mask |= query_mask;
+}
+
+static void d3d12_command_list_add_query_lookup_range(struct d3d12_command_list *list,
+        const struct vkd3d_query_resolve_entry *entry)
+{
+    uint32_t query_index = entry->query_index;
+    uint32_t query_count = entry->query_count;
+    uint32_t advance, bucket, shift;
+    uint64_t query_mask;
+
+    while (query_count)
+    {
+        query_mask = ~0ull;
```

Comment:
This is pretty clever. Would be helpful to have a comment to explain the algorithm. From what I can tell, groups of 64 query indices occupy one hash map entry and the uint64_t bitfield is there to signal individual bits?

Reply:
Yeah, that's how it works. Pretty simple, added a small comment.